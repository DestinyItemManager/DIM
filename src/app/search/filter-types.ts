import { ItemHashTag } from '@destinyitemmanager/dim-api-types';
import { t } from 'app/i18next-t';
import { ItemInfos } from 'app/inventory/dim-item-info';
import { DimItem } from 'app/inventory/item-types';
import { DimStore } from 'app/inventory/store-types';
import { Loadout } from 'app/loadout/loadout-types';
import { InventoryWishListRoll } from 'app/wishlists/wishlists';
type I18nInput = Parameters<typeof t>;

// a filter can return various bool-ish values
type ValidFilterOutput = boolean | null | undefined;

export type ItemFilter = (item: DimItem) => ValidFilterOutput;

/**
 * A slice of data that could be used by filter functions to
 * initialize some data required by particular filters. If a new filter needs
 * context that isn't here, add it to this interface and makeSearchFilterFactory
 * in search-filter.ts.
 */
export interface FilterContext {
  stores: DimStore[];
  allItems: DimItem[];
  currentStore: DimStore;
  loadouts: Loadout[];
  inventoryWishListRolls: { [key: string]: InventoryWishListRoll };
  newItems: Set<string>;
  itemInfos: ItemInfos;
  itemHashTags: {
    [itemHash: string]: ItemHashTag;
  };
  language: string;
}

// TODO: FilterCategory

/**
 * A definition of a filter or closely related group of filters. This is
 * self-contained and can be used for both autocomplete and for building up the
 * filter expression itself. We can also use it to drive filter help and filter
 * editor.
 */
export type FilterDefinition = {
  /**
   * One or more keywords which trigger the filter when typed into search bar.
   * What this means depends on what "format" this filter is.
   */
  keywords: string | string[];

  /**
   * A t()-compatible arg tuple or i18n key pointing to a full description of
   * the filter, to show in filter help
   */
  description: string | I18nInput;

  /**
   * What kind of query this is, used to help generate suggestions.
   *
   * `undefined` - a simple 'is/not' filter. the filter itself knows everything it's looking for
   *
   * `query` - a starting stem and a pre-known value, like "tag:keep". a filterValue will be involved and will match a string we expect
   *
   * `freeform` - a starting stem and a freeform value. the filterValue will be some arbitrary string we test against other strings
   *
   * `range` - a starting stem and a mathlike string afterward like <=5
   *
   * `rangeoverload` - a starting stem like "masterwork" and then either a mathlike string or a word
   *
   * `custom` - suppresses automated suggestion generation so suggestionsGenerator is the only source of suggestions
   */
  format?: 'query' | 'freeform' | 'range' | 'rangeoverload' | 'custom';

  /** destinyVersion - 1 or 2, or if a filter applies to both, undefined */
  destinyVersion?: 1 | 2;

  /** methods for retiring a filter */
  deprecated?: FilterDeprecation;

  /**
   * A function that is given context about the query and the world around it
   * (FilterContext) and should generate a simple filter function that is given
   * an item and returns whether that item should be included in the search.
   * Because this is a function that returns the item filter function, and it is
   * invoked once at the point where we parse the query, it can be used to
   * pre-process information that is needed by the actual filter function. that,
   * given a value from a more complex filter expression and the context about
   * the world around it, can generate a filter function. In that case, the
   * filter function will be generated once, at the point where the overall
   * query is parsed.
   */
  filter: (args: { filterValue: string } & FilterContext) => ItemFilter;

  /**
   * A list of suggested keywords for filters that can take a freeform filter value.
   */
  suggestions?: string[];

  /**
   * A custom function used to generate (additional) suggestions
   */
  // TODO: give this access to a SuggestionsContext arg
  // TODO: add manifest to SuggestionsContext and we can generate archetype/perk/etc suggestions
  suggestionsGenerator?: () => string[];
};

export const enum FilterDeprecation {
  NotDeprecated,
  Deprecated,
  Disabled,
}
