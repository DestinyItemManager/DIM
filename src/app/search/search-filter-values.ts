import { DamageType } from 'bungie-api-ts/destiny2';
import { D1LightStats } from './d1-known-values';
import {
  CUSTOM_TOTAL_STAT_HASH,
  D2ArmorStatHashByName,
  D2LightStats,
  D2WeaponStatHashByName,
  swordStatsByName,
  TOTAL_STAT_HASH,
} from './d2-known-values';

// ✨ magic values ✨
// this file has non-programatically decided information
// hashes, names, & enums, hand-crafted and chosen by us

// this correlation is solely for element filter names
export const damageNamesByEnum: { [key in DamageType]: string | null } = {
  0: null,
  1: 'kinetic',
  2: 'arc',
  3: 'solar',
  4: 'void',
  5: 'raid',
  6: 'stasis',
};

// typescript doesn't understand array.filter
export const damageTypeNames = Object.values(damageNamesByEnum).filter(
  (d) => ![null, 'raid'].includes(d)
) as string[];

/**
 * these stats exist on DIM armor. the 6 real API ones, supplemented by a synthetic Total stat.
 * these are the armor stats that can be looked up by name
 */
export const dimArmorStatHashByName = {
  ...D2ArmorStatHashByName,
  total: TOTAL_STAT_HASH,
  custom: CUSTOM_TOTAL_STAT_HASH,
};

/** stats names used to create armor-specific filters, real ones plus an "any" keyword */
export const searchableArmorStatNames = [...Object.keys(dimArmorStatHashByName), 'any'];

/** armor stat hashes to check for the "any" keyword */
export const armorAnyStatHashes = Object.values(D2ArmorStatHashByName);

/** stat hashes to calculate max values for */
export const armorStatHashes = Object.values(dimArmorStatHashByName);

/** all-stat table, for looking up stat hashes given a queried stat name */
export const statHashByName = {
  ...D2WeaponStatHashByName,
  ...swordStatsByName,
  ...dimArmorStatHashByName,
};

/** all-stat list, to generate filters from */
export const allStatNames = [...Object.keys(statHashByName), 'any'];

export const lightStats = [...D2LightStats, ...D1LightStats];

/** compare against DimItem.type in EN */
export const cosmeticTypes = [
  'Shader',
  'Shaders',
  'Ornaments',
  'Modifications',
  'Emote',
  'Emotes',
  'Emblem',
  'Emblems',
  'Vehicle',
  'Horn',
  'Ship',
  'Ships',
  'ClanBanners',
  'Finishers',
];
