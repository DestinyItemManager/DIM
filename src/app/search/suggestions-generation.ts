import { CustomStatDef } from '@destinyitemmanager/dim-api-types';
import { D2ManifestDefinitions } from 'app/destiny2/d2-definitions';
import { customStatsSelector } from 'app/dim-api/selectors';
import { TagValue } from 'app/inventory/dim-item-info';
import { DimItem } from 'app/inventory/item-types';
import { Loadout } from 'app/loadout-drawer/loadout-types';
import { d2ManifestSelector } from 'app/manifest/selectors';
import { createSelector } from 'reselect';
import {
  allItemsSelector,
  allNotesHashtagsSelector,
  getNotesSelector,
  getTagSelector,
} from '../inventory/selectors';
import { loadoutsSelector } from '../loadout-drawer/selectors';
import { FilterDefinition, SuggestionsContext, canonicalFilterFormats } from './filter-types';

//
// Selectors
//

/**
 * A selector for the suggestionsContext for a particular destiny version.
 * This must depend on every bit of data in suggestionsContext so that we
 * regenerate filter suggestions whenever any of them changes.
 */
export const suggestionsContextSelector = createSelector(
  allItemsSelector,
  loadoutsSelector,
  d2ManifestSelector,
  getTagSelector,
  getNotesSelector,
  allNotesHashtagsSelector,
  customStatsSelector,
  makeSuggestionsContext
);

function makeSuggestionsContext(
  allItems: DimItem[],
  loadouts: Loadout[],
  d2Manifest: D2ManifestDefinitions | undefined,
  getTag: (item: DimItem) => TagValue | undefined,
  getNotes: (item: DimItem) => string | undefined,
  allNotesHashtags: string[],
  customStats: CustomStatDef[]
): SuggestionsContext {
  return {
    allItems,
    loadouts,
    d2Manifest,
    getTag,
    getNotes,
    allNotesHashtags,
    customStats,
  };
}

const operators = ['<', '>', '<=', '>=']; // TODO: add "none"? remove >=, <=?

/**
 * Generates all the possible suggested keywords for the given filter
 *
 * Accepts partial filters with as little as just a "keywords" property,
 * if you want to generate some keywords without a full valid filter
 */
export function generateSuggestionsForFilter(
  filterDefinition: Pick<
    FilterDefinition,
    'keywords' | 'suggestions' | 'format' | 'overload' | 'deprecated'
  >
) {
  return generateGroupedSuggestionsForFilter(filterDefinition, false).flatMap(
    ({ keyword, ops }) => {
      if (ops) {
        return [keyword].concat(ops.map((op) => `${keyword}${op}`));
      } else {
        return [keyword];
      }
    }
  );
}

export function generateGroupedSuggestionsForFilter(
  filterDefinition: Pick<
    FilterDefinition,
    'keywords' | 'suggestions' | 'format' | 'overload' | 'deprecated'
  >,
  forHelp?: boolean
): { keyword: string; ops?: string[] }[] {
  if (filterDefinition.deprecated) {
    return [];
  }

  const { suggestions, keywords } = filterDefinition;
  const thisFilterKeywords = Array.isArray(keywords) ? keywords : [keywords];

  const filterSuggestions = suggestions === undefined ? [] : suggestions;

  const allSuggestions = [];

  const expandFlat = (stringGroups: string[][], minDepth = 0) =>
    expandStringCombinations(stringGroups)
      .slice(minDepth)
      .flat()
      .map((s) => ({ keyword: s }));

  // We delay expanding ops because ops on their own expand the filters list significantly.
  // For autocompletion `generateSuggestionsForFilter` above expands the ops, but the filters
  // help has some special display to group the operator variants.
  const expandOps = (stringGroups: string[][], ops: string[]) => {
    const combinations = expandStringCombinations([...stringGroups, []]);
    const partialSuggestions = combinations
      .slice(0, stringGroups.length - 1)
      .flat()
      .map((s) => ({ keyword: s }));
    const opSuggestions = combinations[stringGroups.length - 1].map((s) => ({ keyword: s, ops }));
    return partialSuggestions.concat(opSuggestions);
  };

  for (const format of canonicalFilterFormats(filterDefinition.format)) {
    switch (format) {
      case 'simple':
        // Pass minDepth 1 to not generate "is:" and "not:" suggestions. Only generate `is:` for filters help
        allSuggestions.push(
          ...expandFlat([forHelp ? ['is'] : ['is', 'not'], thisFilterKeywords], 1)
        );
        break;
      case 'query':
        // `query` is exhaustive, so only include keyword: for autocompletion, not filters help
        allSuggestions.push(
          ...expandFlat([thisFilterKeywords, filterSuggestions], forHelp ? 1 : 0)
        );
        break;
      case 'freeform':
        allSuggestions.push(...expandFlat([thisFilterKeywords, []]));
        break;
      case 'range':
        allSuggestions.push(...expandOps([thisFilterKeywords], operators));
        if (filterDefinition.overload) {
          const overloadNames = Object.keys(filterDefinition.overload);
          allSuggestions.push(...expandFlat([thisFilterKeywords, overloadNames], 1));

          // Outside of filter help (i.e. only for autocompletion), also expand overloaded ranges like season:<current
          if (!forHelp) {
            allSuggestions.push(
              ...expandFlat(
                [
                  thisFilterKeywords,
                  operators.flatMap((op) =>
                    overloadNames.map((overloadName) => `${op}${overloadName}`)
                  ),
                ],
                1
              )
            );
          }
        }
        break;
      case 'stat':
        // stat lists aren't exhaustive
        allSuggestions.push(...expandOps([thisFilterKeywords, filterSuggestions], operators));
        break;
      case 'custom':
        break;
    }
  }

  return allSuggestions;
}

/**
 * loops through collections of strings (filter segments),
 * generating combinations grouped by number of segments
 *
 * for example, with
 * `[ [a], [b,c], [d,e] ]`
 * as an input, this generates
 *
 * `[ [a:], [a:b:, a:c:], [a:b:d, a:b:e, a:c:d, a:c:e] ]`
 */
function expandStringCombinations(stringGroups: string[][]) {
  const results: string[][] = [];
  for (let i = 0; i < stringGroups.length; i++) {
    const stringGroup = stringGroups[i];
    const stems = results.length ? results[results.length - 1] : undefined;
    const newResults = stringGroup.flatMap((suffix) =>
      stems
        ? stems.map(
            (stem) =>
              (stem ? `${stem}${suffix}` : suffix) + (i === stringGroups.length - 1 ? '' : ':')
          )
        : [`${suffix}:`]
    );
    results.push(newResults);
  }
  return results;
}
