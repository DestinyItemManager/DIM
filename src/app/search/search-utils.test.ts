import { FilterDefinition } from './filter-types';
import { allStatNames, searchableArmorStatNames } from './search-filter-values';
import { generateSuggestionsForFilter } from './suggestions-generation';

describe('generateSuggestionsForFilter', () => {
  const cases: [
    format: FilterDefinition['format'],
    keywords: FilterDefinition['keywords'],
    suggestions: FilterDefinition['suggestions'],
    overload: { [key: string]: number } | undefined
  ][] = [
    [undefined, ['a', 'b', 'c'], undefined, undefined],
    ['query', 'a', ['b', 'c'], undefined],
    ['stat', 'a', ['b', 'c'], undefined],
    ['range', 'a', undefined, undefined],
    ['range', 'a', undefined, { worthy: 10, arrivals: 11 }],
    ['freeform', 'a', ['b', 'c'], undefined],
    [undefined, ['a'], undefined, undefined],
    ['stat', 'stat', allStatNames, undefined],
    ['query', 'maxstatvalue', searchableArmorStatNames, undefined],
    ['query', 'maxstatvalue', searchableArmorStatNames, undefined],
    ['range', 'energycapacity', undefined, undefined],
  ];

  test.each(cases)(
    "full suggestions for filter format '%s', keyword '%s' with suggestions %s, overload %s",
    (
      format: FilterDefinition['format'],
      keywords: string | string[],
      suggestions?: string[],
      overload?: { [key: string]: number } | undefined
    ) => {
      const candidates = generateSuggestionsForFilter({
        format,
        keywords,
        suggestions,
        overload,
      });
      expect(candidates).toMatchSnapshot();
    }
  );
});
