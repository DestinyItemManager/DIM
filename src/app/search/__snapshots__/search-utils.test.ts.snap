// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateSuggestionsForFilter full suggestions for filter format 'freeform', keyword 'a' with suggestions [ 'b', 'c' ] 1`] = `
Array [
  "a:",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'query', keyword 'a' with suggestions [ 'b', 'c' ] 1`] = `
Array [
  "a:",
  "a:b",
  "a:c",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'query', keyword 'maxstatvalue' with suggestions [
  'mobility',  'resilience',
  'recovery',  'discipline',
  'intellect', 'strength',
  'total',     'custom',
  'any'
] 1`] = `
Array [
  "maxstatvalue:",
  "maxstatvalue:mobility",
  "maxstatvalue:resilience",
  "maxstatvalue:recovery",
  "maxstatvalue:discipline",
  "maxstatvalue:intellect",
  "maxstatvalue:strength",
  "maxstatvalue:total",
  "maxstatvalue:custom",
  "maxstatvalue:any",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'query', keyword 'maxstatvalue' with suggestions [
  'mobility',  'resilience',
  'recovery',  'discipline',
  'intellect', 'strength',
  'total',     'custom',
  'any'
] 2`] = `
Array [
  "maxstatvalue:",
  "maxstatvalue:mobility",
  "maxstatvalue:resilience",
  "maxstatvalue:recovery",
  "maxstatvalue:discipline",
  "maxstatvalue:intellect",
  "maxstatvalue:strength",
  "maxstatvalue:total",
  "maxstatvalue:custom",
  "maxstatvalue:any",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'range', keyword 'a' with suggestions [ 'b', 'c' ] 1`] = `
Array [
  "a:",
  "a:b:",
  "a:c:",
  "a:b:<",
  "a:c:<",
  "a:b:>",
  "a:c:>",
  "a:b:<=",
  "a:c:<=",
  "a:b:>=",
  "a:c:>=",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'range', keyword 'a' with suggestions undefined 1`] = `
Array [
  "a:",
  "a:<",
  "a:>",
  "a:<=",
  "a:>=",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'range', keyword 'stat' with suggestions [
  'rpm',             'rof',
  'charge',          'impact',
  'handling',        'range',
  'stability',       'reload',
  'magazine',        'aimassist',
  'equipspeed',      'velocity',
  'blastradius',     'recoildirection',
  'drawtime',        'zoom',
  'inventorysize',   'swingspeed',
  'guardefficiency', 'guardresistance',
  'chargerate',      'guardendurance',
  'ammocapacity',    'mobility',
  'resilience',      'recovery',
  'discipline',      'intellect',
  'strength',        'total',
  'custom',          'any'
] 1`] = `
Array [
  "stat:",
  "stat:rpm:",
  "stat:rof:",
  "stat:charge:",
  "stat:impact:",
  "stat:handling:",
  "stat:range:",
  "stat:stability:",
  "stat:reload:",
  "stat:magazine:",
  "stat:aimassist:",
  "stat:equipspeed:",
  "stat:velocity:",
  "stat:blastradius:",
  "stat:recoildirection:",
  "stat:drawtime:",
  "stat:zoom:",
  "stat:inventorysize:",
  "stat:swingspeed:",
  "stat:guardefficiency:",
  "stat:guardresistance:",
  "stat:chargerate:",
  "stat:guardendurance:",
  "stat:ammocapacity:",
  "stat:mobility:",
  "stat:resilience:",
  "stat:recovery:",
  "stat:discipline:",
  "stat:intellect:",
  "stat:strength:",
  "stat:total:",
  "stat:custom:",
  "stat:any:",
  "stat:rpm:<",
  "stat:rof:<",
  "stat:charge:<",
  "stat:impact:<",
  "stat:handling:<",
  "stat:range:<",
  "stat:stability:<",
  "stat:reload:<",
  "stat:magazine:<",
  "stat:aimassist:<",
  "stat:equipspeed:<",
  "stat:velocity:<",
  "stat:blastradius:<",
  "stat:recoildirection:<",
  "stat:drawtime:<",
  "stat:zoom:<",
  "stat:inventorysize:<",
  "stat:swingspeed:<",
  "stat:guardefficiency:<",
  "stat:guardresistance:<",
  "stat:chargerate:<",
  "stat:guardendurance:<",
  "stat:ammocapacity:<",
  "stat:mobility:<",
  "stat:resilience:<",
  "stat:recovery:<",
  "stat:discipline:<",
  "stat:intellect:<",
  "stat:strength:<",
  "stat:total:<",
  "stat:custom:<",
  "stat:any:<",
  "stat:rpm:>",
  "stat:rof:>",
  "stat:charge:>",
  "stat:impact:>",
  "stat:handling:>",
  "stat:range:>",
  "stat:stability:>",
  "stat:reload:>",
  "stat:magazine:>",
  "stat:aimassist:>",
  "stat:equipspeed:>",
  "stat:velocity:>",
  "stat:blastradius:>",
  "stat:recoildirection:>",
  "stat:drawtime:>",
  "stat:zoom:>",
  "stat:inventorysize:>",
  "stat:swingspeed:>",
  "stat:guardefficiency:>",
  "stat:guardresistance:>",
  "stat:chargerate:>",
  "stat:guardendurance:>",
  "stat:ammocapacity:>",
  "stat:mobility:>",
  "stat:resilience:>",
  "stat:recovery:>",
  "stat:discipline:>",
  "stat:intellect:>",
  "stat:strength:>",
  "stat:total:>",
  "stat:custom:>",
  "stat:any:>",
  "stat:rpm:<=",
  "stat:rof:<=",
  "stat:charge:<=",
  "stat:impact:<=",
  "stat:handling:<=",
  "stat:range:<=",
  "stat:stability:<=",
  "stat:reload:<=",
  "stat:magazine:<=",
  "stat:aimassist:<=",
  "stat:equipspeed:<=",
  "stat:velocity:<=",
  "stat:blastradius:<=",
  "stat:recoildirection:<=",
  "stat:drawtime:<=",
  "stat:zoom:<=",
  "stat:inventorysize:<=",
  "stat:swingspeed:<=",
  "stat:guardefficiency:<=",
  "stat:guardresistance:<=",
  "stat:chargerate:<=",
  "stat:guardendurance:<=",
  "stat:ammocapacity:<=",
  "stat:mobility:<=",
  "stat:resilience:<=",
  "stat:recovery:<=",
  "stat:discipline:<=",
  "stat:intellect:<=",
  "stat:strength:<=",
  "stat:total:<=",
  "stat:custom:<=",
  "stat:any:<=",
  "stat:rpm:>=",
  "stat:rof:>=",
  "stat:charge:>=",
  "stat:impact:>=",
  "stat:handling:>=",
  "stat:range:>=",
  "stat:stability:>=",
  "stat:reload:>=",
  "stat:magazine:>=",
  "stat:aimassist:>=",
  "stat:equipspeed:>=",
  "stat:velocity:>=",
  "stat:blastradius:>=",
  "stat:recoildirection:>=",
  "stat:drawtime:>=",
  "stat:zoom:>=",
  "stat:inventorysize:>=",
  "stat:swingspeed:>=",
  "stat:guardefficiency:>=",
  "stat:guardresistance:>=",
  "stat:chargerate:>=",
  "stat:guardendurance:>=",
  "stat:ammocapacity:>=",
  "stat:mobility:>=",
  "stat:resilience:>=",
  "stat:recovery:>=",
  "stat:discipline:>=",
  "stat:intellect:>=",
  "stat:strength:>=",
  "stat:total:>=",
  "stat:custom:>=",
  "stat:any:>=",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'rangeoverload', keyword 'a' with suggestions [ 'b', 'c' ] 1`] = `
Array [
  "a:",
  "a:<",
  "a:>",
  "a:<=",
  "a:>=",
  "a:",
  "a:b",
  "a:c",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'rangeoverload', keyword 'energycapacity' with suggestions [ 'any', 'arc', 'solar', 'void', 'ghost', 'subclass' ] 1`] = `
Array [
  "energycapacity:",
  "energycapacity:<",
  "energycapacity:>",
  "energycapacity:<=",
  "energycapacity:>=",
  "energycapacity:",
  "energycapacity:any",
  "energycapacity:arc",
  "energycapacity:solar",
  "energycapacity:void",
  "energycapacity:ghost",
  "energycapacity:subclass",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'undefined', keyword '[ 'a' ]' with suggestions undefined 1`] = `
Array [
  "is:a",
  "not:a",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'undefined', keyword '[ 'a', 'b', 'c' ]' with suggestions undefined 1`] = `
Array [
  "is:a",
  "not:a",
  "is:b",
  "not:b",
  "is:c",
  "not:c",
]
`;
